### PR TITLE
Implementation of geofencing functionality

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds geofencing capabilities to QField configured via qfieldsync. The geofencing functionality features three behavior that can:
- alert users when within areas (default behavior)
- alert users when outside all areas
- inform users when entering and leaving an area

The configuration in qfieldsync is done so through the project properties dialog's qfieldsync panel:

image

On the QField side, opening a project will automatically adopt the geofencing settings. The picked polygon layer's display name setting will be used as area name.

UX when geofencing is set to alert users when entering geofenced areas:

https://github.com/opengisch/QField/assets/1728657/fcb169d2-5442-414c-a8e0-a5a4060b8e64

UX when geofencing is set to alert users when outside all geofenced areas:

https://github.com/opengisch/QField/assets/1728657/da1a7ce6-98b7-4598-8a12-b739dc0a6450

UX when geofencing is set to inform users when entering/leaving geofenced areas:

https://github.com/opengisch/QField/assets/1728657/225ba369-4a40-404c-89e1-1ca5de490953

Finally, a new option found in the settings' positioning tab allows for users to prevent digitizing while in geofencing 'alert' mode:

image

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5e00b4d00268ce7694a8428f55ea8253ab4936e3: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5409-19666b90bb0aebbeedc69ece7465a4af926ff630-android-x86.apk)
</details>

File: src/qml/qgismobileapp.qml [hunk 1]
```diff
@@ -861,6 +862,82 @@ ApplicationWindow {
...
+    height: width
+    radius: width / 2
+    visible: false
+
+    x: parent.width - width / 2
+    y: locationToolbar.y + gnssButton.y + (gnssButton.height / 2) - height / 2
+
+    color: Theme.errorColor
+    border.width: 2
+    border.color: "#ffffff"
```
Comment (mohsenD98) [inline src/qml/qgismobileapp.qml:938]: I'm not sure, can we use Theme.color ?

**Review by mohsenD98 — APPROVED** [review]
Fantastic feature! Very promising. really enjoyed it. 
I just put some ideas and cleanups on `qml` files.

File: src/qml/qgismobileapp.qml [hunk 2]
```diff
@@ -861,6 +862,82 @@ ApplicationWindow {
...
+      } else if (behavior == Geofencer.AlertWhenOutsideGeofencedArea && !geofencer.isWithin) {
+        platformUtilities.vibrate(500)
+        displayToast(qsTr("Position outside areas after leaving ‘%1’").arg(lastWithinAreaName), 'error')
+      } else if (behavior == Geofencer.InformWhenEnteringLeavingGeofencedArea) {
+        if (isWithin) {
+          platformUtilities.vibrate(250)
+          displayToast(qsTr("Position entered into ‘%1’").arg(isWithinAreaName))
+        } else if (lastWithinAreaName != '') {
+          platformUtilities.vibrate(250)
+          displayToast(qsTr("Position left from ‘%1’").arg(lastWithinAreaName))
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/qml/qgismobileapp.qml:884]: @mohsenD98 , I don't think that's super appropriate to show in logs, it'll end up abusing the alert icon and making people ignore that symbol. We really want to reserve that for crucial warnings/errors.

File: src/qml/qgismobileapp.qml (hunk 1 above)
Comment (nirvn) [reply src/qml/qgismobileapp.qml:938]: Good catch, this was only use to tweak location, removed.

File: src/qml/qgismobileapp.qml [hunk 3]
```diff
@@ -861,6 +862,82 @@ ApplicationWindow {
...
 
+  Geofencer {
+    id: geofencer
+
+    position: positionSource.projectedPosition
+    positionCrs: mapCanvas.mapSettings.destinationCrs
+
+    onIsWithinChanged: {
+      if (behavior == Geofencer.AlertWhenInsideGeofencedArea && geofencer.isWithin) {
+        platformUtilities.vibrate(500)
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/qml/qgismobileapp.qml:873]: Done.

**Pull request merged**